File: app/src/main/java/de/tum/in/tumcampusapp/component/other/generic/DrawerMenuHelper.java
package de.tum.in.tumcampusapp.component.other.generic;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.widget.DrawerLayout;
import android.view.Menu;
import android.view.MenuItem;
import android.view.SubMenu;

import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.api.tumonline.AccessTokenManager;
import de.tum.in.tumcampusapp.component.other.settings.UserPreferencesActivity;
import de.tum.in.tumcampusapp.component.tumui.calendar.CalendarActivity;
import de.tum.in.tumcampusapp.component.tumui.grades.GradesActivity;
import de.tum.in.tumcampusapp.component.tumui.lectures.activity.LecturesPersonalActivity;
import de.tum.in.tumcampusapp.component.tumui.person.PersonsSearchActivity;
import de.tum.in.tumcampusapp.component.tumui.roomfinder.RoomFinderActivity;
import de.tum.in.tumcampusapp.component.tumui.tutionfees.TuitionFeesActivity;
import de.tum.in.tumcampusapp.component.ui.barrierfree.BarrierFreeInfoActivity;
import de.tum.in.tumcampusapp.component.ui.cafeteria.activity.CafeteriaActivity;
import de.tum.in.tumcampusapp.component.ui.chat.activity.ChatRoomsActivity;
import de.tum.in.tumcampusapp.component.ui.curricula.CurriculaActivity;
import de.tum.in.tumcampusapp.component.ui.news.NewsActivity;
import de.tum.in.tumcampusapp.component.ui.openinghour.OpeningHoursListActivity;
import de.tum.in.tumcampusapp.component.ui.plan.PlansActivity;
import de.tum.in.tumcampusapp.component.ui.studyroom.StudyRoomsActivity;
import de.tum.in.tumcampusapp.component.ui.transportation.TransportationActivity;
import de.tum.in.tumcampusapp.utils.Const;
import de.tum.in.tumcampusapp.utils.Utils;

public class DrawerMenuHelper implements NavigationView.OnNavigationItemSelectedListener {

    private static final SideNavigationItem[] MY_TUM = {
            new SideNavigationItem(R.string.schedule, R.drawable.ic_calendar, CalendarActivity.class, true, false),
            new SideNavigationItem(R.string.my_lectures, R.drawable.ic_my_lectures, LecturesPersonalActivity.class, true, false),
            new SideNavigationItem(R.string.chat_rooms, R.drawable.ic_comment, ChatRoomsActivity.class, true, true),
            new SideNavigationItem(R.string.my_grades, R.drawable.ic_my_grades, GradesActivity.class, true, false),
            new SideNavigationItem(R.string.tuition_fees, R.drawable.ic_money, TuitionFeesActivity.class, true, false),
            };

    private static final SideNavigationItem[] COMMON_TUM = {
            new SideNavigationItem(R.string.menues, R.drawable.ic_cutlery, CafeteriaActivity.class, false, false),
            new SideNavigationItem(R.string.study_rooms, R.drawable.ic_group_work, StudyRoomsActivity.class, false, false),
            new SideNavigationItem(R.string.roomfinder, R.drawable.ic_place, RoomFinderActivity.class, false, false),
            new SideNavigationItem(R.string.plans, R.drawable.ic_plans, PlansActivity.class, false, false),
            new SideNavigationItem(R.string.mvv, R.drawable.ic_mvv, TransportationActivity.class, false, false),
            new SideNavigationItem(R.string.person_search, R.drawable.ic_users, PersonsSearchActivity.class, true, false),
            new SideNavigationItem(R.string.news, R.drawable.ic_rss, NewsActivity.class, false, false),
            new SideNavigationItem(R.string.barrier_free, R.drawable.ic_accessible, BarrierFreeInfoActivity.class, false, false),
            new SideNavigationItem(R.string.opening_hours, R.drawable.ic_time, OpeningHoursListActivity.class, false, false),
            new SideNavigationItem(R.string.study_plans, R.drawable.ic_study_plans, CurriculaActivity.class, false, false)

    };

    private static final SideNavigationItem[] APP = {
            new SideNavigationItem(R.string.settings, R.drawable.ic_action_settings, UserPreferencesActivity.class, false, false)
    };

    private final Context mContext;

    private final DrawerLayout mDrawerLayout;

    public DrawerMenuHelper(Context context, DrawerLayout drawerLayout) {
        mContext = context;
        mDrawerLayout = drawerLayout;
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
        mDrawerLayout.closeDrawers();
        Intent intent = menuItem.getIntent();
        intent.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION);
        mContext.startActivity(intent);
        return true;
    }

    public void populateMenu(Menu navigationMenu) {
        boolean hasTUMOAccess = new AccessTokenManager(mContext).hasValidAccessToken();
        boolean chatEnabled = Utils.getSettingBool(mContext, Const.GROUP_CHAT_ENABLED, false);

        if (hasTUMOAccess) {
            SubMenu myTumMenu = navigationMenu.addSubMenu(R.string.my_tum);
            for (SideNavigationItem item : MY_TUM) {
                if (!(item.needsChatAccess && !chatEnabled)) {
                    myTumMenu.add(item.titleRes)
                             .setIcon(item.iconRes)
                             .setIntent(new Intent(mContext, item.activity));
                }
            }
        }

        SubMenu commonTumMenu = navigationMenu.addSubMenu(R.string.tum_common);
        for (SideNavigationItem item : COMMON_TUM) {
            if (!(item.needsTUMOAccess && !hasTUMOAccess)) {
                commonTumMenu.add(item.titleRes)
                             .setIcon(item.iconRes)
                             .setIntent(new Intent(mContext, item.activity));
            }
        }
        for (SideNavigationItem item : APP) {
            navigationMenu.add(item.titleRes)
                          .setIcon(item.iconRes)
                          .setIntent(new Intent(mContext, item.activity));
        }
    }

    private static class SideNavigationItem {
        final int titleRes;
        final int iconRes;
        final Class<? extends Activity> activity;
        final boolean needsTUMOAccess;
        final boolean needsChatAccess;

        SideNavigationItem(int titleRes, int iconRes, Class<? extends Activity> activity, boolean needsTUMOAccess, boolean needsChatAccess) {
            this.titleRes = titleRes;
            this.iconRes = iconRes;
            this.activity = activity;
            this.needsTUMOAccess = needsTUMOAccess;
            this.needsChatAccess = needsChatAccess;
        }
    }
}

File: app/src/main/java/de/tum/in/tumcampusapp/component/other/generic/drawer/DrawerMenuHelper.kt
package de.tum.`in`.tumcampusapp.component.other.generic.drawer

import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import com.google.android.material.navigation.NavigationView
import de.tum.`in`.tumcampusapp.R
import de.tum.`in`.tumcampusapp.api.tumonline.AccessTokenManager
import de.tum.`in`.tumcampusapp.component.other.settings.SettingsActivity
import de.tum.`in`.tumcampusapp.component.tumui.calendar.CalendarFragment
import de.tum.`in`.tumcampusapp.component.tumui.feedback.FeedbackActivity
import de.tum.`in`.tumcampusapp.component.tumui.grades.GradesFragment
import de.tum.`in`.tumcampusapp.component.tumui.lectures.fragment.LecturesFragment
import de.tum.`in`.tumcampusapp.component.tumui.person.PersonSearchFragment
import de.tum.`in`.tumcampusapp.component.tumui.roomfinder.RoomFinderFragment
import de.tum.`in`.tumcampusapp.component.tumui.tutionfees.TuitionFeesFragment
import de.tum.`in`.tumcampusapp.component.ui.barrierfree.BarrierFreeInfoFragment
import de.tum.`in`.tumcampusapp.component.ui.cafeteria.fragment.CafeteriaFragment
import de.tum.`in`.tumcampusapp.component.ui.chat.ChatRoomsFragment
import de.tum.`in`.tumcampusapp.component.ui.news.NewsFragment
import de.tum.`in`.tumcampusapp.component.ui.openinghour.OpeningHoursListFragment
import de.tum.`in`.tumcampusapp.component.ui.overview.InformationActivity
import de.tum.`in`.tumcampusapp.component.ui.overview.MainFragment
import de.tum.`in`.tumcampusapp.component.ui.studyroom.StudyRoomsFragment
import de.tum.`in`.tumcampusapp.component.ui.ticket.activity.EventsFragment
import de.tum.`in`.tumcampusapp.utils.Const
import de.tum.`in`.tumcampusapp.utils.Utils
import de.tum.`in`.tumcampusapp.utils.allItems
import de.tum.`in`.tumcampusapp.utils.plusAssign

class DrawerMenuHelper(
    private val activity: AppCompatActivity,
    private val navigationView: NavigationView
) {

    private val currentFragment: Fragment?
        get() = activity.supportFragmentManager.findFragmentById(R.id.contentFrame)

    private val navigationMenu: Menu
        get() = navigationView.menu

    private val allItems = mutableListOf<NavItem>()

    fun populateMenu() {
        val hasTumOnlineAccess = AccessTokenManager.hasValidAccessToken(activity)
        val isChatEnabled = Utils.getSettingBool(activity, Const.GROUP_CHAT_ENABLED, false)
        val isEmployeeMode = Utils.getSettingBool(activity, Const.EMPLOYEE_MODE, false)

        navigationMenu.clear()
        allItems.clear()

        navigationMenu += HOME
        allItems += HOME

        val myTumMenu = navigationMenu.addSubMenu(R.string.my_tum)
        if (hasTumOnlineAccess) {
            val candidates = MY_TUM
                .filterNot { !isChatEnabled && it.needsChatAccess }
                .filterNot { isEmployeeMode && it.hideForEmployees }

            for (candidate in candidates) {
                myTumMenu += candidate
                allItems += candidate
            }
        }

        val generalMenu = navigationMenu.addSubMenu(R.string.common_info)
        val generalCandidates = GENERAL.filterNot { it.needsTUMOAccess && !hasTumOnlineAccess }
        for (candidate in generalCandidates) {
            generalMenu += candidate
            allItems += candidate
        }

        val aboutMenu = navigationMenu.addSubMenu(R.string.about)
        for (item in ABOUT) {
            aboutMenu += item
            allItems += item
        }

        highlightCurrentItem()
    }

    fun findNavItem(menuItem: MenuItem): NavItem {
        if (menuItem.title == activity.getString(HOME.titleRes)) {
            return HOME
        }

        for (item in MY_TUM + GENERAL) {
            if (menuItem.title == activity.getString(item.titleRes)) {
                return item
            }
        }

        for (item in ABOUT) {
            if (menuItem.title == activity.getString(item.titleRes)) {
                return item
            }
        }

        throw IllegalArgumentException("Invalid menu item ${menuItem.title} provided")
    }

    fun updateNavDrawer() {
        highlightCurrentItem()
    }

    private fun highlightCurrentItem() {
        val items = navigationMenu.allItems
        items.forEach { it.isChecked = false }

        val currentIndex = allItems
            .mapNotNull { it as? NavItem.FragmentDestination }
            .indexOfFirst { it.fragment == currentFragment?.javaClass }

        if (currentIndex != -1) {
            items[currentIndex].isCheckable = true
            items[currentIndex].isChecked = true
        }
    }

    private companion object {

        private val HOME = NavItem.FragmentDestination(R.string.home, R.drawable.ic_outline_home_24px, MainFragment::class.java)

        private val MY_TUM = arrayOf(
                NavItem.FragmentDestination(R.string.calendar, R.drawable.ic_outline_event_24px, CalendarFragment::class.java, true),
                NavItem.FragmentDestination(R.string.my_lectures, R.drawable.ic_outline_school_24px, LecturesFragment::class.java, true, hideForEmployees = true),
                NavItem.FragmentDestination(R.string.chat_rooms, R.drawable.ic_outline_chat_bubble_outline_24px, ChatRoomsFragment::class.java, true, true),
                NavItem.FragmentDestination(R.string.my_grades, R.drawable.ic_outline_insert_chart_outlined_24px, GradesFragment::class.java, true, hideForEmployees = true),
                NavItem.FragmentDestination(R.string.tuition_fees, R.drawable.ic_money, TuitionFeesFragment::class.java, true, hideForEmployees = true)
        )

        private val GENERAL = arrayOf(
                NavItem.FragmentDestination(R.string.menues, R.drawable.ic_cutlery, CafeteriaFragment::class.java),
                NavItem.FragmentDestination(R.string.study_rooms, R.drawable.ic_outline_group_work_24px, StudyRoomsFragment::class.java),
                NavItem.FragmentDestination(R.string.roomfinder, R.drawable.ic_outline_location_on_24px, RoomFinderFragment::class.java),
                NavItem.FragmentDestination(R.string.person_search, R.drawable.ic_outline_people_outline_24px, PersonSearchFragment::class.java, true),
                NavItem.FragmentDestination(R.string.news, R.drawable.ic_rss, NewsFragment::class.java),
                NavItem.FragmentDestination(R.string.events_tickets, R.drawable.tickets, EventsFragment::class.java),
                NavItem.FragmentDestination(R.string.barrier_free, R.drawable.ic_outline_accessible_24px, BarrierFreeInfoFragment::class.java),
                NavItem.FragmentDestination(R.string.opening_hours, R.drawable.ic_outline_access_time_24px, OpeningHoursListFragment::class.java)
        )

        private val ABOUT = arrayOf(
                NavItem.ActivityDestination(R.string.show_feedback, R.drawable.ic_outline_feedback_24px, FeedbackActivity::class.java),
                NavItem.ActivityDestination(R.string.about_tca, R.drawable.ic_action_info_black, InformationActivity::class.java),
                NavItem.ActivityDestination(R.string.settings, R.drawable.ic_outline_settings_24px, SettingsActivity::class.java)
        )
    }
}
